refinery/tool_manager/utils.py

```diff
@@ -11,3 +17,132 @@ class AdminFieldPopulator(admin.ModelAdmin):
+                # the bottom-most nesting.
+                fr_store.append(file_relationship)
+                # NOTE: Recursive functions need to return themselves,
+                # otherwise Python returns `None`
+                return create_nesting(
+                    value, fr_store=fr_store)
+            else:
+                # If we reach here, we have reached the bottom-most nested
+                # file_relationship. Since we want to act upon the bottom-most
+                # file_relationship's input_files, we can safely grab the
+                # last element from our fr_store due to Python's nature or
+                # ordering lists.
+                bottom_file_relationship = fr_store[-1]
+
+                # Fetch, create, and associate InputFiles with the
+                # bottom-most file_relationship
+                if d["input_files"]:
+                    for input_file in d["input_files"]:
+
+                        input_file_instance = InputFileFactory(
+                            name=input_file["name"],
+                            description=input_file["description"])
+
+                        allowed_filetypes = input_file["allowed_filetypes"]
+
+                        for key in allowed_filetypes:
+                            # NOTE: we are not handling the usual .get()
+                            # exceptions here due to the transaction
+                            # management wrapping
+                            # create_tool_definition_from_workflow()
+                            file_type = FileType.objects.get(name=key["name"])
+
+                            input_file_instance.allowed_filetypes.add(
+                                file_type)
+                            bottom_file_relationship.input_files.add(
+                                input_file_instance)
+
+                # Iterate through stored FileRelationship objects and
+                # associate their children
+                for idx, f in enumerate(fr_store):
+                    try:
+                        f.file_relationship.add(fr_store[idx + 1])
+                    except IndexError:
+                        # Return the top-most FileRelationship
+                        return fr_store[0]
+
+
+def validate_workflow_annotation(wf_dict):
+    """
+    Validate incoming annotation data to ensure ToolDefinitions are created
+    properly.
+    :param annotation: dict containing Galaxy Workflow annotation data
+    :return: Boolean: True if validation passes.
+    """
+    keys_to_check = [
+        "file_relationship", "description", "output_files"]
+    keys_not_found = []
+
+    for key in keys_to_check:
+        if key not in wf_dict["annotation"]:
+            keys_not_found.append(key)
+
+    if keys_not_found:
+        raise CommandError(
+            "Workflow not properly Annotated. Keys: {} were not found in "
+            "workflow annotation data.".format(keys_not_found))
+    else:
+        return True
```

Comment:
Its expected since atm `CommandError` will stop execution, but I'l just have to handle the `CommandError` in the mgmt command itself so this will change